api/src/main/java/org/openmrs/api/impl/CohortServiceImpl.java

```diff
@@ -57,108 +60,146 @@ public Cohort saveCohort(Cohort cohort) throws APIException {
 		if (log.isInfoEnabled()) {
 			log.info("Saving cohort " + cohort);
 		}
-		
+
 		return dao.saveCohort(cohort);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohort(java.lang.Integer)
 	 */
 	@Transactional(readOnly = true)
 	public Cohort getCohort(Integer id) {
 		return dao.getCohort(id);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#voidCohort(org.openmrs.Cohort, java.lang.String)
 	 */
 	public Cohort voidCohort(Cohort cohort, String reason) {
 		// other setters done by the save handlers
 		return Context.getCohortService().saveCohort(cohort);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#getCohortByUuid(java.lang.String)
 	 */
 	@Transactional(readOnly = true)
 	public Cohort getCohortByUuid(String uuid) {
 		return dao.getCohortByUuid(uuid);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#addPatientToCohort(org.openmrs.Cohort,
 	 *      org.openmrs.Patient)
 	 */
 	public Cohort addPatientToCohort(Cohort cohort, Patient patient) {
 		if (!cohort.contains(patient)) {
-			cohort.getMemberIds().add(patient.getPatientId());
+			CohortMembership cohortMembership = new CohortMembership(patient);
+            Context.getCohortService().addMembershipToCohort(cohort, cohortMembership);
 			Context.getCohortService().saveCohort(cohort);
 		}
 		return cohort;
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.CohortService#removePatientFromCohort(org.openmrs.Cohort,
 	 *      org.openmrs.Patient)
 	 */
 	public Cohort removePatientFromCohort(Cohort cohort, Patient patient) {
 		if (cohort.contains(patient)) {
```

Comment:
Based on your implementation of equals, this isn't doing the correct thing, a patient can have multiple memberships in the same cohort although technically there should be one active one. You want to remove the active membership for this patient and not just any membership